src/main/java/org/opensearch/plugin/insights/rules/transport/top_queries/TransportTopQueriesAction.java

```diff
@@ -68,27 +75,143 @@ public TransportTopQueriesAction(
         this.queryInsightsService = queryInsightsService;
     }
 
+    ActionListener<TopQueriesResponse> createInMemoryDataCollectionListener(
+        TopQueriesRequest request,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        return new ActionListener<TopQueriesResponse>() {
+            @Override
+            public void onResponse(TopQueriesResponse inMemoryQueriesResponse) {
+                handleInMemoryDataResponse(request, inMemoryQueriesResponse, finalListener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                finalListener.onFailure(e);
+            }
+        };
+    }
+
+    void handleInMemoryDataResponse(
+        TopQueriesRequest request,
+        TopQueriesResponse inMemoryQueriesResponse,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        List<TopQueries> inMemoryTopQueries = inMemoryQueriesResponse.getNodes();
+        List<FailedNodeException> inMemoryDataFailures = inMemoryQueriesResponse.failures();
+        String from = request.getFrom();
+        String to = request.getTo();
+        if (from != null && to != null) {
+            fetchHistoricalData(request, inMemoryTopQueries, inMemoryDataFailures, finalListener);
```

Comment:
Why we couldn't have directly called `fetchHistoricalData` when this condition suffice? As collecting `inMemory` data doesn't much sense in such cases.